streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -254,17 +254,37 @@ protected <R> QueryResult<R> runRangeQuery(final Query<R> query,
         final QueryResult<R> result;
         final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
         final RangeQuery<Bytes, byte[]> rawRangeQuery;
+        final boolean isKeyAscending = typedQuery.isKeyAscending();
         if (typedQuery.getLowerBound().isPresent() && typedQuery.getUpperBound().isPresent()) {
-            rawRangeQuery = RangeQuery.withRange(
-                keyBytes(typedQuery.getLowerBound().get()),
-                keyBytes(typedQuery.getUpperBound().get())
-            );
+            if (isKeyAscending) {
+                rawRangeQuery = RangeQuery.withRange(
+                        keyBytes(typedQuery.getLowerBound().get()),
+                        keyBytes(typedQuery.getUpperBound().get())
+                );
+            } else {
+                rawRangeQuery = (RangeQuery<Bytes, byte[]>) (RangeQuery<?, ?>) RangeQuery.withRange(
+                        keyBytes(typedQuery.getLowerBound().get()),
+                        keyBytes(typedQuery.getUpperBound().get())
+                ).withDescendingKeys();
```

Comment:
Was looking into this. It's a type erasure issue. You need to specify the type expliclity for this case when calling the static method so avoid the casts:
```
rawRangeQuery = RangeQuery.<Bytes, byte[]>withRange(
    keyBytes(typedQuery.getLowerBound().get()),
    keyBytes(typedQuery.getUpperBound().get())
).withDescendingKeys();
```

Reply:
However, it seem code is a little bit redundant here. I think it would make sense to lift the `final` for `rawRangeQuery` and do the following:

```
rawRangeQuery = RangeQuery.withRange(
    keyBytes(typedQuery.getLowerBound().get()),
    keyBytes(typedQuery.getUpperBound().get())
);
if (!isKeyAscending)
    rawRangeQuery = rawRangeQuery.withDescendingKeys();
}
```

Reply:
Thank you, I choose  lift the final for rawRangeQuery and have already updated it.